telegram/_messageentity.py

```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
 
         return super().de_json(data=data, bot=bot)
 
+    @staticmethod
+    def adjust_message_entities_to_utf_16(
+        text: str, entities: Sequence["MessageEntity"]
+    ) -> Sequence["MessageEntity"]:
+        """Utility functionality for converting the offset and length of entities from
+        Unicode (`str`) to UTF-16 (`utf-16-le` encoded `bytes`).
+
+        Tip:
+            Only the offsets and lengths calulated in UTF-16 is acceptable by the Telegram Bot API.
+            If they are calculated using the Unicode string (`str` object), errors may occur
+            when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+            For more information, see `Unicode <https://en.wikipedia.org/wiki/Unicode>` and
+            `Plane (Unicode) <https://en.wikipedia.org/wiki/Plane_(Unicode)>`.
+
+        Examples:
+            Below is a snippet of code that demonstrates how to use this function to convert
+            entities from Unicode to UTF-16 space. The `unicode_entities` are calculated in
+            Unicode and the `utf_16_entities` are calculated in UTF-16.
+
+        .. code-block:: python
+
+            text = "𠌕 bold 𝄢 italic underlined: 𝛙𝌢𑁍"
+            unicode_entities = [
+                MessageEntity(offset=2, length=4, type=MessageEntity.BOLD),
+                MessageEntity(offset=9, length=6, type=MessageEntity.ITALIC),
+                MessageEntity(offset=28, length=3, type=MessageEntity.UNDERLINE),
+            ]
+            utf_16_entities = MessageEntity.adjust_message_entities_to_utf_16(
+                text, unicode_entities
+            )
+            await bot.send_message(
+                chat_id=123,
+                text=text,
+                entities=utf_16_entities,
+            )
+            # utf_16_entities[0]: offset=3, length=4
+            # utf_16_entities[1]: offset=11, length=6
+            # utf_16_entities[2]: offset=30, length=6
+
+        Args:
+            text (:obj:`str`): The text that the entities belong to
+            entities (List[:class:`telegram.MessageEntity`]): Sequence of entities
+                with offset and length calculated in Unicode
+
+        Returns:
+            Sequence[:class:`telegram.MessageEntity`]: Sequence of entities
+            with offset and length calculated in UTF-16 encoding
+        """
+        # get sorted offsets
+        offsets_list = sorted(
+            itertools.chain(*((x.offset, x.offset + x.length) for x in entities))
+        )
+        accumulated_length = 0
+        according_utf16_lengths: List[int] = []
```

Comment:
I see no need for this helper variable. You can directly write the caculated postiions into "~offset~position_translation" as explained in https://github.com/python-telegram-bot/python-telegram-bot/pull/4323#discussion_r1655410569